In [32]:
from pyspark.sql import SparkSession, functions as F, Window

spark = SparkSession.builder \
    .appName("Footbal project") \
    .getOrCreate()

results_df = spark.read.option("header", "true").csv("results.csv")
goals_df = spark.read.option("header", "true").csv("goalscorers.csv")

results_df.show(5)
goals_df.show(5)


+----------+---------+---------+----------+----------+----------+-------+--------+-------+
|      date|home_team|away_team|home_score|away_score|tournament|   city| country|neutral|
+----------+---------+---------+----------+----------+----------+-------+--------+-------+
|1872-11-30| Scotland|  England|         0|         0|  Friendly|Glasgow|Scotland|  FALSE|
|1873-03-08|  England| Scotland|         4|         2|  Friendly| London| England|  FALSE|
|1874-03-07| Scotland|  England|         2|         1|  Friendly|Glasgow|Scotland|  FALSE|
|1875-03-06|  England| Scotland|         2|         2|  Friendly| London| England|  FALSE|
|1876-03-04| Scotland|  England|         3|         0|  Friendly|Glasgow|Scotland|  FALSE|
+----------+---------+---------+----------+----------+----------+-------+--------+-------+
only showing top 5 rows
+----------+---------+---------+---------+----------------+------+--------+-------+
|      date|home_team|away_team|     team|          scorer|minute|own_goa

Prvi zadatak: najbolja i najgora gol razlika Srbije 

In [ ]:
rdd = results_df.rdd.map(lambda row: (
    row['home_team'], 
    row['away_team'], 
    int(row['home_score']), 
    int(row['away_score'])
))

serbia_rdd = rdd.filter(lambda x: x[0] == 'Serbia' or x[1] == 'Serbia')

def map_goal_diff(match):
    home, away, home_score, away_score = match
    if home == 'Serbia':
        opponent = away
        diff = home_score - away_score
    else:
        opponent = home
        diff = away_score - home_score
    return (opponent, diff)

mapped = serbia_rdd.map(map_goal_diff)
reduced = mapped.reduceByKey(lambda a, b: a + b)

best = reduced.max(key=lambda x: x[1])
worst = reduced.min(key=lambda x: x[1])

print("Najbolja gol razlika sa Srbijom:", best)
print("Najgora gol razlika sa Srbijom:", worst)

Najbolja gol razlika:
Najgora gol razlika:

Drugi zadatak: Fudbaleri sa het-trikom na Evropskom

In [ ]:
df2 = goals_df.join(results_df, on=["date", "home_team", "away_team"], how="left")

euro_df = df2.filter(F.col("tournament") == "UEFA Euro")

hat_tricks = (
    euro_df.groupBy("date", "home_team", "away_team", "team", "scorer", "tournament", "country")
    .agg(F.count("*").alias("golova"))
    .filter(F.col("golova") >= 3)
)

hat_tricks.select(
    F.col("date").alias("Datum"),
    F.col("home_team").alias("Domaći"),
    F.col("away_team").alias("Gosti"),
    F.col("team").alias("Igra za"),
    F.col("scorer").alias("Strelac"),
    F.col("golova").alias("Golova na utakmici"),
    F.col("tournament").alias("Turnir"),
    F.col("country").alias("Zemlja")
).show(truncate=False)

+----------+-----------+-----------+-----------+----------------+------------------+---------+-----------+
|Datum     |Domaći     |Gosti      |Igra za    |Strelac         |Golova na utakmici|Turnir   |Zemlja     |
+----------+-----------+-----------+-----------+----------------+------------------+---------+-----------+
|1984-06-19|France     |Yugoslavia |France     |Michel Platini  |3                 |UEFA Euro|France     |
|1984-06-16|France     |Belgium    |France     |Michel Platini  |3                 |UEFA Euro|France     |
|2000-06-25|Netherlands|Serbia     |Netherlands|Patrick Kluivert|3                 |UEFA Euro|Netherlands|
|1976-06-17|Yugoslavia |Germany    |Germany    |Dieter Müller   |3                 |UEFA Euro|Yugoslavia |
|2000-06-20|Portugal   |Germany    |Portugal   |Sérgio Conceição|3                 |UEFA Euro|Netherlands|
|1988-06-15|England    |Netherlands|Netherlands|Marco van Basten|3                 |UEFA Euro|Germany    |
|1980-06-14|Germany    |Netherlands|G

+----------+-----------+-----------+-----------+----------------+------------------+---------+-----------+
|Datum     |Domaći     |Gosti      |Igra za    |Strelac         |Golova na utakmici|Turnir   |Zemlja     |
+----------+-----------+-----------+-----------+----------------+------------------+---------+-----------+
|1984-06-19|France     |Yugoslavia |France     |Michel Platini  |3                 |UEFA Euro|France     |
|1984-06-16|France     |Belgium    |France     |Michel Platini  |3                 |UEFA Euro|France     |
|2000-06-25|Netherlands|Serbia     |Netherlands|Patrick Kluivert|3                 |UEFA Euro|Netherlands|
|1976-06-17|Yugoslavia |Germany    |Germany    |Dieter Müller   |3                 |UEFA Euro|Yugoslavia |
|2000-06-20|Portugal   |Germany    |Portugal   |Sérgio Conceição|3                 |UEFA Euro|Netherlands|
|1988-06-15|England    |Netherlands|Netherlands|Marco van Basten|3                 |UEFA Euro|Germany    |
|1980-06-14|Germany    |Netherlands|Germany    |Klaus Allofs    |3                 |UEFA Euro|Italy      |
|2008-06-10|Spain      |Russia     |Spain      |David Villa     |3                 |UEFA Euro|Austria    |
+----------+-----------+-----------+-----------+----------------+------------------+---------+-----------+


Treci zadatak: Najbolji strelci na svetskom od 1990.

In [31]:
df3 = goals_df.join(results_df, on=["date", "home_team", "away_team"], how="left")

df_wc = df3.filter(F.col("tournament") == "FIFA World Cup") \
        .withColumn("godina", F.year(F.to_date(F.col("date"), "yyyy-MM-dd"))) \
        .filter(F.col("godina") >= 1990)

scorers = df_wc.groupBy("scorer", "team", "godina") \
        .agg(F.count("*").alias("Golova"),
             F.first("tournament").alias("Turnir"),
             F.first("country").alias("Zemlja"))

window = Window.partitionBy("godina").orderBy(F.desc("Golova"))

top_scorers = scorers.withColumn("rang", F.row_number().over(window)) \
                     .filter(F.col("rang") <= 2) \
                     .orderBy("godina", F.desc("Golova"))


top_scorers.select(
    F.col("scorer").alias("Strelac"),
    F.col("team").alias("Reprezentacija"),
    F.col("Golova"),
    F.col("Rang"),
    F.col("Godina"),
    F.col("Turnir"),
    F.col("Zemlja")
).show(truncate=False)

+-------------------+--------------+------+----+------+--------------+-------------+
|Strelac            |Reprezentacija|Golova|Rang|Godina|Turnir        |Zemlja       |
+-------------------+--------------+------+----+------+--------------+-------------+
|Salvatore Schillaci|Italy         |6     |1   |1990  |FIFA World Cup|Italy        |
|Tomáš Skuhravý     |Czechoslovakia|5     |2   |1990  |FIFA World Cup|Italy        |
|Hristo Stoichkov   |Bulgaria      |6     |1   |1994  |FIFA World Cup|United States|
|Oleg Salenko       |Russia        |6     |2   |1994  |FIFA World Cup|United States|
|Davor Šuker        |Croatia       |6     |1   |1998  |FIFA World Cup|France       |
|Christian Vieri    |Italy         |5     |2   |1998  |FIFA World Cup|France       |
|Ronaldo            |Brazil        |8     |1   |2002  |FIFA World Cup|South Korea  |
|Miroslav Klose     |Germany       |5     |2   |2002  |FIFA World Cup|Japan        |
|Miroslav Klose     |Germany       |5     |1   |2006  |FIFA World

+-------------------+--------------+------+----+------+--------------+-------------+
|Strelac            |Reprezentacija|Golova|Rang|Godina|Turnir        |Zemlja       |
+-------------------+--------------+------+----+------+--------------+-------------+
|Salvatore Schillaci|Italy         |6     |1   |1990  |FIFA World Cup|Italy        |
|Tomáš Skuhravý     |Czechoslovakia|5     |2   |1990  |FIFA World Cup|Italy        |
|Hristo Stoichkov   |Bulgaria      |6     |1   |1994  |FIFA World Cup|United States|
|Oleg Salenko       |Russia        |6     |2   |1994  |FIFA World Cup|United States|
|Davor Šuker        |Croatia       |6     |1   |1998  |FIFA World Cup|France       |
|Christian Vieri    |Italy         |5     |2   |1998  |FIFA World Cup|France       |
|Ronaldo            |Brazil        |8     |1   |2002  |FIFA World Cup|South Korea  |
|Miroslav Klose     |Germany       |5     |2   |2002  |FIFA World Cup|Japan        |
|Miroslav Klose     |Germany       |5     |1   |2006  |FIFA World Cup|Germany      |
|David Villa        |Spain         |3     |2   |2006  |FIFA World Cup|Germany      |
|David Villa        |Spain         |5     |1   |2010  |FIFA World Cup|South Africa |
|Diego Forlán       |Uruguay       |5     |2   |2010  |FIFA World Cup|South Africa |
|James Rodríguez    |Colombia      |6     |1   |2014  |FIFA World Cup|Brazil       |
|Thomas Müller      |Germany       |5     |2   |2014  |FIFA World Cup|Brazil       |
|Harry Kane         |England       |6     |1   |2018  |FIFA World Cup|Russia       |
|Antoine Griezmann  |France        |4     |2   |2018  |FIFA World Cup|Russia       |
|Kylian Mbappé      |France        |8     |1   |2022  |FIFA World Cup|Qatar        |
|Lionel Messi       |Argentina     |7     |2   |2022  |FIFA World Cup|Qatar        |
+-------------------+--------------+------+----+------+--------------+-------------+


Cetvrti zadatak: 20 reprezentacija koji su najduze bez pobede od 1980.

In [38]:
wc_matches = results_df.filter(F.col("tournament") == "FIFA World Cup")
df4 = results_df.withColumn("date", F.to_date(F.col("date"), "yyyy-MM-dd"))
df4 = df4.filter(F.year("date") >= 1980)

wc_teams = (wc_matches.select(F.col("home_team").alias("team"))
            .union(wc_matches.select(F.col("away_team").alias("team")))
            .distinct())

team_list = [row['team'] for row in wc_teams.collect()]

losses = (df4.filter(F.col("home_score") < F.col("away_score"))
            .select(F.col("home_team").alias("team"), "date", F.col("away_team").alias("opponent")))
losses = losses.union(
    df4.filter(F.col("away_score") < F.col("home_score"))
      .select(F.col("away_team").alias("team"), "date", F.col("home_team").alias("opponent"))
)
losses = losses.orderBy("team", "date")

w = Window.partitionBy("team").orderBy("date")
losses = (losses.withColumn("Prethodni_poraz_datum", F.lag("date").over(w))
                .withColumn("Prethodni_poraz_od", F.lag("opponent").over(w))
                .withColumn("Period_dana", F.datediff(F.col("date"), F.col("Prethodni_poraz_datum"))))

w2 = Window.partitionBy("team").orderBy(F.desc("Period_dana"))
losses = losses.withColumn("rang", F.row_number().over(w2)).filter(F.col("rang") == 1)

(losses.select(
    "team", 
    "opponent", 
    "date", 
    "Prethodni_poraz_od", 
    "Prethodni_poraz_datum", 
    "Period_dana")
          .withColumnRenamed("team", "Reprezentacija")
          .withColumnRenamed("date", "Poraz_datum")
          .withColumnRenamed("opponent", "Poraz od")
          .filter(F.col("Reprezentacija").isin(team_list))
          .orderBy(F.desc("Period_dana"))
          .limit(20)).show(truncate=False)


+-------------------+-------------------+-----------+------------------+---------------------+-----------+
|Reprezentacija     |Poraz od           |Poraz_datum|Prethodni_poraz_od|Prethodni_poraz_datum|Period_dana|
+-------------------+-------------------+-----------+------------------+---------------------+-----------+
|Yugoslavia         |Basque Country     |1997-12-26 |Netherlands       |1992-03-25           |2102       |
|North Korea        |Kuwait             |1998-09-13 |South Korea       |1993-10-28           |1781       |
|Cuba               |Trinidad and Tobago|1992-06-24 |Guatemala         |1988-04-30           |1516       |
|Haiti              |Colombia           |1989-06-27 |Guatemala         |1985-05-15           |1504       |
|Mexico             |Canada             |1990-05-13 |England           |1986-05-17           |1457       |
|United States      |Guatemala          |1984-10-14 |Mexico            |1980-11-09           |1435       |
|DR Congo           |Ivory Coast     

+-------------------+-------------------+-----------+------------------+---------------------+-----------+
|Reprezentacija     |Poraz od           |Poraz_datum|Prethodni_poraz_od|Prethodni_poraz_datum|Period_dana|
+-------------------+-------------------+-----------+------------------+---------------------+-----------+
|Yugoslavia         |Basque Country     |1997-12-26 |Netherlands       |1992-03-25           |2102       |
|North Korea        |Kuwait             |1998-09-13 |South Korea       |1993-10-28           |1781       |
|Cuba               |Trinidad and Tobago|1992-06-24 |Guatemala         |1988-04-30           |1516       |
|Haiti              |Colombia           |1989-06-27 |Guatemala         |1985-05-15           |1504       |
|Mexico             |Canada             |1990-05-13 |England           |1986-05-17           |1457       |
|United States      |Guatemala          |1984-10-14 |Mexico            |1980-11-09           |1435       |
|DR Congo           |Ivory Coast        |1985-03-05 |Ghana             |1981-08-02           |1311       |
|Spain              |France             |1998-01-28 |Italy             |1994-07-09           |1299       |
|Argentina          |Saudi Arabia       |2022-11-22 |Brazil            |2019-07-02           |1239       |
|Panama             |Cuba               |1984-03-18 |Honduras          |1980-12-14           |1190       |
|Algeria            |Equatorial Guinea  |2022-01-16 |Benin             |2018-10-16           |1188       |
|Trinidad and Tobago|Canada             |1988-10-02 |Costa Rica        |1985-07-18           |1172       |
|Iraq               |China PR           |2000-09-03 |Kazakhstan        |1997-06-29           |1162       |
|Italy              |Spain              |2021-10-06 |Portugal          |2018-09-10           |1122       |
|France             |Denmark            |1996-11-09 |Bulgaria          |1993-11-17           |1088       |
|Israel             |Northern Ireland   |1984-10-16 |Northern Ireland  |1981-11-18           |1063       |
|Saudi Arabia       |Egypt              |1992-09-17 |South Korea       |1989-10-25           |1058       |
|Ivory Coast        |Mali               |1983-01-09 |Egypt             |1980-03-08           |1037       |
|Costa Rica         |United States      |1988-06-14 |Honduras          |1985-09-08           |1010       |
|Colombia           |Bolivia            |1994-04-07 |Argentina         |1991-07-21           |991        |
+-------------------+-------------------+-----------+------------------+---------------------+-----------+